client/client.go

```diff
@@ -37,21 +41,97 @@ func New(logID int64, cc *grpc.ClientConn) *LogClient {
 	}
 }
 
-// AddLeaf adds leaf to the append only log. It blocks until a verifiable response is received.
+// AddLeaf adds leaf to the append only log.
+// Blocks until it gets a verifiable response.
 func (c *LogClient) AddLeaf(ctx context.Context, data []byte) error {
+	leaf := buildLeaf(data)
+	if err := c.queueLeaf(ctx, leaf); err != nil {
+		return err
+	}
+
+	// Wait for TreeSize to update.
+	if err := c.waitForSTRUpdate(ctx, 2); err != nil {
+		return nil
+	}
+
+	// Get proof by hash.
+	err := c.getInclusionProof(ctx, leaf.MerkleLeafHash, c.STR.TreeSize)
+	if err != nil {
+		return err
+		// TODO(gdbelvin): retry with backoff and jitter
+	}
+
+	return err
+}
+
+// waitForSTRUpdate repeatedly fetches the STR until the TreeSize changes
+// or until a maximum number of attempts (10) has been tried.
+func (c *LogClient) waitForSTRUpdate(ctx context.Context, maxRetries int) error {
+	startTreeSize := c.STR.TreeSize
+	b := &backoff.Backoff{
+		Min:    100 * time.Millisecond,
+		Max:    10 * time.Second,
+		Factor: 2,
+		Jitter: true,
+	}
+	for i := 0; startTreeSize == c.STR.TreeSize; i++ {
+		time.Sleep(b.Duration())
+		if err := c.UpdateSTR(ctx); err != nil {
+			return err
+		}
+		if i >= maxRetries {
```

Comment:
I think this should be factored into the stop condition of the loop.

We've got 2 issues here:
* The last attempt will always cause an error, regardless of a successful update
* The loop does maxRetries+1 attempts, not maxRetries

Please add tests for the scenarios above before changing it.